File: 01/005/Main.java
public class Main {

    public static boolean allReducable(long num, long reducer) {
        for (long i = 1; i <= reducer; i++) {
            if (num % i != 0) {
                return false;
            }
        }
        return true;
    }
    
    public static long reduceResult(long num, long reducer) {
        for (long i = reducer; i > 1; i--) {
            if (allReducable(num / i, reducer)) {
                return reduceResult(num / i, reducer);
            }
        }
        return num;
    }
    
    public static long smallestDivisible(long reducer) {
        long result = 1;
        for (long i = 2; i <= reducer; i++) {
            result = reduceResult(result * i, i);
        }
        return result;
    }
    
    public static void main(String[] args) {
        System.out.println(smallestDivisible(20));
    }
}

File: 01/005/main.rb
def allReducable num, reducer
    for i in 1..reducer do
        if num % i != 0 then
            return false
        end
    end
    true
end

def reduceResult num, reducer
    for i in (2..reducer).reverse_each do
        if allReducable(num/i, reducer) then
            return reduceResult(num/i, reducer)
        end
    end
    num
end

def smallestDivisible reducer
    result = 1
    for i in 2..reducer do
        result = reduceResult(result*i, i)
    end
    result
end

puts smallestDivisible(20)
